torch_glow/src/CachingGraphRunner.cpp

```diff
@@ -20,61 +20,128 @@
-                                         torch::jit::Stack &stack) {
-  // If this is the first time this subgraph has been run then create a new
-  // GraphInfo object to store information about it.
+namespace {
+/// Returns a hash that represents a given PT subgraph represented by the PT
+/// node \p node and a set of tensor shape inputs to that subgraph from the \p
+/// stack.
+size_t getGraphHash(const torch::jit::Node *node,
+                    const torch::jit::Stack &stack) {
+  const std::shared_ptr<torch::jit::Graph> graph = node->g(at::attr::Subgraph);
+  const auto &graphInputs = graph->inputs();
+  const auto inputs = torch::jit::last(stack, graphInputs.size());
+  torch::jit::CompleteArgumentSpec spec(/*with_grad*/ false, inputs);
+
+  return torch::hash_combine(reinterpret_cast<size_t>(node), spec.hashCode());
+}
+
+} // namespace
 
+llvm::Expected<CachingGraphRunner::PerGlowGraphInfo *>
+CachingGraphRunner::loadGraphImpl(const torch::jit::Node *node,
+                                  torch::jit::Stack &stack) {
   const std::shared_ptr<torch::jit::Graph> graph = node->g(at::attr::Subgraph);
+  RETURN_ERR_IF_NOT(graph, "A subgraph couldn't be found for this node");
+
   const auto &graphInputs = graph->inputs();
-  const auto numInputs = graphInputs.size();
-  auto inputs = torch::jit::last(stack, numInputs);
-  const char *const functionName = "PyTorchFunction";
+  auto inputs = torch::jit::last(stack, graphInputs.size());
+
+  size_t hash = getGraphHash(node, stack);
+
+  // If we already have a Glow function compiled for this graph with and the
+  // given inputs then use that.
+  auto it = perGlowGraphInfoMap.find(hash);
+  if (it != perGlowGraphInfoMap.end()) {
+    return it->second.get();
+  }
+
+  auto info = llvm::make_unique<PerGlowGraphInfo>();
+  info->functionName = strFormat("PTFunction%lu", hash);
+  info->hash = hash;
+  info->node = node;
 
-  glow::Function *f = nullptr;
-  executionEngine_.setBackendName(executionEngine_.getBackendName());
-  auto &mod = executionEngine_.getModule();
-  f = mod.createFunction(functionName);
-  std::vector<glow::Placeholder *> inputPlaceholders;
-  std::vector<glow::Placeholder *> outputPlaceholders;
+  std::unique_ptr<Module> module = llvm::make_unique<Module>();
+  Function *f = module->createFunction(info->functionName);
 
   RETURN_IF_ERR(PyTorchModelLoader::loadJITGraph(
-      *f, *graph, inputs, inputPlaceholders, outputPlaceholders,
+      *f, *graph, inputs, info->inputPlaceholders, info->outputPlaceholders,
       getPyTorchLoaderSettings()));
 
   glow::CompilationContext cctx;
-  executionEngine_.compile(cctx);
 
-  glow::PlaceholderBindings bindings;
-  for (size_t i = 0; i < inputs.size(); ++i) {
-    glow::Placeholder *ph = inputPlaceholders[i];
+  RETURN_IF_ERR(hostManager_->addNetwork(std::move(module), cctx));
+
+  perGlowGraphInfoMap.insert({hash, std::move(info)});
+
+  return perGlowGraphInfoMap[hash].get();
```

Comment:
Searching again, use the return iterator at line 78.

Reply:
Even better you can insert dummy info:
  `auto p = perGlowGraphInfoMap.insert({hash, PerGlowGraphInfo()});
   if (!p.second) {
     return p.first->second;
   }

   populate p.first->second with correct values.
    

   `

Reply:
I did it this way because in case of failure there isn't a dummy node floating in perGlowGraphInfoMap that has to be removed, the extra lookup is fine imo since this only happens the first time when we're compiling the glow function which will be much slower anyways

Reply:
It can be done more elegantly, using llvm::scope_exit

   auto p = perGlowGraphInfoMap.insert({hash, PerGlowGraphInfo()});
   auto guard =  llvm::make_scope_exit([&]() {
     perGlowGraphInfoMap.erase(p.first);
   });
  
  // on failure
  return error;

   // on success
  guard.release();
  return success;

Reply:
I don't really prefer that way, I'd rather just not insert into the map until we are sure that it's good